bootstrap_salt/deploy_lib/github.py

```diff
@@ -0,0 +1,218 @@
+    teams = get_teams(org)
+    team = filter(lambda x: x['slug'] == team_slug, teams)
+    if len(team) == 0:
+        raise InvalidTeamException(
+            'Team {} is not part of org {}'.format(team_slug, org))
+    else:
+        return team[0]
+
+
+def get_team_members(team_slug, org=ORG):
+    try:
+        team = get_org_team(team_slug, org)
+    except InvalidTeamException:
+        return {}
+    url = '{}teams/{}/members'.format(API_ENDPOINT, team['id'])
+    return get_paginated_content(url)
+
+
+def check_org_membership(org, user):
+    url = '{}orgs/{}/members/{}'.format(API_ENDPOINT, org, user)
+    r = requests.get(url, auth=(get_github_token(), 'x-oauth-basic'))
+    if r.status_code == 204:
+        return True
+    else:
+        return False
+
+
+def get_key_fingerprint(key):
+    toks = key['key'].split()
+    try:
+        (enc, key_body, descr) = key['key'].split()
+    except ValueError:
+        if len(toks) == 2:
+            (enc, key_body) = key['key'].split()
+        else:
+            raise InvalidKey('Invalid key: {}'.format(key['key']))
+    if enc not in ['ssh-rsa', 'ssh-dss', 'ssh-ecdsa']:
+        raise InvalidKey('Invalid enc type: {}'.format(enc))
+    key_bytes = base64.b64decode(key_body)
+    md5hash = hashlib.md5()
+    md5hash.update(key_bytes)
+    fingerprint = md5hash.hexdigest()
+    fingerprint = [fingerprint[i:i+2] for i in range(0, len(fingerprint), 2)]
+    return ':'.join(fingerprint)
+
+
+def get_user_keys(user, fingerprints=None):
+    keys = get_paginated_content('{}users/{}/keys'.format(API_ENDPOINT, user))
+    if fingerprints:
+        for key in keys:
+            fingerprint = get_key_fingerprint(key)
+            if fingerprint not in fingerprints:
+                keys.remove(key)
+    return keys
+
+
+def get_user_struct(username, user_data=None):
+    user_data = user_data if user_data else {}
+    print 'Adding key for {}'.format(username)
+    fingerprints = user_data.get('fingerprints', None)
+    keys = get_user_keys(username, fingerprints=fingerprints)
+    username = user_data.get('unix_username', username)
+    fields = ('enc', 'key')
+    key_tuples = (key['key'].split(' ') for key in keys)
+    key_dict = [dict(zip(fields, map(str, x))) for x in key_tuples]
+    # The three lines above can be written in the form:
+    # [dict(zip(('enc', 'key'), key.split(' '))) for key in keys]
+    # ... but good luck figuring it out in a week from now
+    if not key_dict:
+        return {}
+    return {str(username): {'public_keys': key_dict}}
+
+
+def handle_string_team(team, org):
+    team_users = get_team_members(team, org)
+    return team_users, []
+
+
+def handle_dict_team(team, org):
+    team_name = team.keys()[0]
+    special_users = team[team_name]
+    team_users = get_team_members(team_name, org)
+    return team_users, special_users
+
+
+def get_keys(data):
+    out = {}
+    for org in data.keys():
+        for team in data[org]['teams']:
```

Comment:
```
Traceback (most recent call last):
  File "/Users/aberlin/.virtualenvs/docker-registry-deploy/lib/python2.7/site-packages/fabric/main.py", line 743, in main
    *args, **kwargs
  File "/Users/aberlin/.virtualenvs/docker-registry-deploy/lib/python2.7/site-packages/fabric/tasks.py", line 424, in execute
    results['<local-only>'] = task.run(*args, **new_kwargs)
  File "/Users/aberlin/.virtualenvs/docker-registry-deploy/lib/python2.7/site-packages/fabric/tasks.py", line 174, in run
    return self.wrapped(*args, **kwargs)
  File "/Users/aberlin/code/moj/bootstrap-salt/bootstrap_salt/fab_tasks.py", line 217, in generate_ssh_key_pillar
    ssh_key_data = github.get_keys(key_config)
  File "/Users/aberlin/code/moj/bootstrap-salt/bootstrap_salt/deploy_lib/github.py", line 157, in get_keys
    for team in data[org]['teams']:
KeyError: 'teams'
```

Or if we're making this required make the error say where teams key needs to live